AWS: documentation page for AWS module

jackye1995 opened the pull request:
Initial draft for documentation page of the AWS module, I will mark this as draft until #1823 and #1844 are merged.

Also, I am currently placing it under `Tables` tab, which is clearly wrong. I was thinking about adding a new tab `Integrations` for all integrations including AWS and Nessie, but the navbar becomes too long. I am thinking about moving Flink, Hive and Presto all to a single `Engines` tab, any thoughts on that?

ismailsimsek commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+When this configurer is used, a STS client is initialized with default [credentials chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html) and [region chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html),
+and all the other clients (Glue, DynamoDB, S3, etc.) will use the configured assume-role credential and region.
+
+## Run Iceberg on AWS
```
Thank you @jackye1995  it looks great, this is more an question. is it also possible to run iceberg with AWS "Glue Job"? if we use `spark.jars.packages` to provide iceberg library.

jackye1995 replied:
I have not tried that yet since I am mostly focused on Athena use case, but let me check if anyone has done that, thanks for bringing up this use case.

yyanyy commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+## Run Iceberg on AWS
+
+[Amazon EMR](https://aws.amazon.com/emr/) is the most common platform to run Iceberg on AWS. EMR can provision clusters with [Spark](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark.html) (EMR 6 for Spark 3, EMR 5 for Spark 2),
```
Minor: did you have to pull down the aws sdk bundle to run sparksql in EMR? IIRC there's already some preinstalled aws libraries on EMR, if that's true we might be able to mention them here to save one step for the users.

jackye1995 replied:
The v2 library is not bundled in spark.

HeartSaVioR commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+To store data in a different local or cloud store, Glue catalog can switch to use `HadoopFileIO` 
+or any custom FileIO using the mechanism described in the [custom FileIO](../custom-catalog/#custom-file-io-implementation) section.
+
+## S3 FileIO
```
I was about to ask the rationalization of S3 FileIO compared to Hadoop filesystem API with S3 support in #1945, but this section covers it. Thanks!

Probably worth to also mention whether Hadoop FS API with S3 is sufficient to work with, or S3 FileIO is required to avoid consistency glitches. That would help end users to determine whether including aws module is a kind of requirement for dealing with S3 or not.

jackye1995 replied:
I have updated strong consistency section for more details, please let me know if it is enough.

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+## S3 FileIO
+
+Iceberg allows user to write data to S3 through `S3FileIO`.
```
```suggestion
Iceberg allows users to write data to S3 through `S3FileIO`.
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+This provides maximized upload speed and minimized local disk usage during uploads.
+Here are the configurations user can tune related to this feature:
+
+* `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared pool across all output streams)
```
Is there a default here? Or is there no default and setting this is required?

jackye1995 replied:
default to `Runtime.getRuntime().availableProcessors()`, let me add in the doc

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+Here are the configurations user can tune related to this feature:
+
+* `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared pool across all output streams)
+* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, default to 32MB
```
```suggestion
* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, defaults to 32MB
```
or
```suggestion
* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, default is 32MB
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+* `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared pool across all output streams)
+* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, default to 32MB
+* `s3fileio.multipart.threshold`: the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload, default to 1.5
```
as above `defaults to` or `default is`

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+* `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared pool across all output streams)
+* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, default to 32MB
+* `s3fileio.multipart.threshold`: the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload, default to 1.5
+* `s3fileio.staging.dir`: the directory to hold temporary files, default to Java's `java.io.tmpdir` property value
```
as above to/is

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+To enable server side encryption, use the following configuration properties:
+
+* `s3fileio.sse.type`: `none`, `s3`, `kms` or `custom`, default to `none`
```
As above for defaults.

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+S3 and many other cloud storage services [throttle requests based on object prefix](https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/). 
+This means data stored in a traditional Hive storage layout has bad read and write throughput since data files of the same partition are placed under the same prefix.
+Iceberg by default uses the Hive storage layout, but can be switch to use a different `ObjectStoreLocationProvider`.
```
```suggestion
Iceberg by default uses the Hive storage layout, but can be configured to use a different `ObjectStoreLocationProvider`.
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+Iceberg by default uses the Hive storage layout, but can be switch to use a different `ObjectStoreLocationProvider`.
+In this mode, a hash string is added to the beginning of each file path, so that files are equally distributed across all prefixes in an S3 bucket.
+This results in minimized throttling and maximized throughput for S3-related IO operations.
+For more details, please follow [LocationProvider Configuration](../custom-catalog/#custom-location-provider-implementation) section to enable this feature.  
```
```suggestion
For more details, please refer to the [LocationProvider Configuration](../custom-catalog/#custom-location-provider-implementation) section to see how to enable this feature.  
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+### S3 Strong Consistency
+
+In November 2020, S3 announced [strong consistency](https://aws.amazon.com/s3/consistency/) for all GET and LIST operations, and Iceberg is updated to fully leverage this feature.
+When creating a new output file using `OutputFile.create()`, strong consistency check is used and an `AlreadyExistsException` will be thrown if the file already exists in S3.
```
```suggestion
When creating a new output file using `OutputFile.create()`, a strong consistency check is used and an `AlreadyExistsException` will be thrown if the file already exists in S3.
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+Before `S3FileIO` was introduced, many Iceberg users choose to use `HadoopFileIO` to write data to S3 through the [S3A FileSystem](https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java).
+As introduced in the above sections, the `S3FileIO` adopts latest AWS clients and S3 features for optimzied security and performance,
+ and is thus recommend for S3 use case over S3A.
```
```suggestion
 and is thus recommend for S3 use cases rather than the S3A FileSystem.
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+In November 2020, S3 announced [strong consistency](https://aws.amazon.com/s3/consistency/) for all GET and LIST operations, and Iceberg is updated to fully leverage this feature.
+When creating a new output file using `OutputFile.create()`, strong consistency check is used and an `AlreadyExistsException` will be thrown if the file already exists in S3.
+
+### Hadoop S3A
```
```suggestion
### Hadoop S3A FileSystem
```
I think it's useful to refer to the FileSystem otherwise some people might get confused by the differences between S3, S3A, S3N etc.

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+`S3FileIO` is compatible with legacy URI schemes written by S3A, 
+so any existing tables with `s3a://` or `s3n://` file paths are treated as equivalent `s3://` file paths.    
+
+If for any reason you have to use S3A, here are the instructions:
```
Here you are talking about storing data using `s3a://` URLs as opposed to the above discussion about the S3A FileSystem right? This should probably be reworded to be clear about that if that is the case.

jackye1995 replied:
Thank you, I updated the section, please let me know if it is clearer.

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+If for any reason you have to use S3A, here are the instructions:
+
+1. to store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 
```
```suggestion
1. To store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+If for any reason you have to use S3A, here are the instructions:
+
+1. to store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 
+2. For `HiveCatalog`, to store metadata also using S3A, specify Hadoop config `hive.metastore.warehouse.dir` to be an S3A path.
```
```suggestion
2. For `HiveCatalog`, to also store metadata using S3A, specify the Hadoop config property `hive.metastore.warehouse.dir` to be an S3A path.
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+1. to store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 
+2. For `HiveCatalog`, to store metadata also using S3A, specify Hadoop config `hive.metastore.warehouse.dir` to be an S3A path.
+3. Add [hadoop-aws](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-aws) as a runtime dependency, configure AWS settings based on [hadoop-aws documentation](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html) (make sure you check the version, s3a configuration varies a lot based on the version you use)   
```
```suggestion
3. Add [hadoop-aws](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-aws) as a runtime dependency, configure AWS settings based on [hadoop-aws documentation](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html) (make sure you check the version as S3A configuration varies a lot based on the version you use) .
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+## AWS client configurations
+
+Many organizations have customized their way of obtaining AWS credential and region information and configuring details about AWS clients for features like proxy access, retry, etc.
+Therefore, we open a configurer interface for Iceberg users to plugin any client configuration in a centralized place. 
```
```suggestion
Therefore, we open a configurer interface for Iceberg users to plug in any client configuration in a centralized place. 
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+* `client.assume-role.external-id`: optional external ID for the role to assume
+* `client.assume-role.region`: a region for all clients (except the STS client) to use
+
+When this configurer is used, a STS client is initialized with default [credentials chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html) and [region chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html),
```
```suggestion
When this configurer is used, a STS client is initialized with a default [credentials chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html) and [region chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html),
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+[Hive](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-hive.html), [Flink](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-flink.html),
+[Presto](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-presto.html) that can run Iceberg.
+
+[Amazon Kinesis Data Analytics](https://aws.amazon.com/about-aws/whats-new/2019/11/you-can-now-run-fully-managed-apache-flink-applications-with-apache-kafka/) also provdies a way 
```
```suggestion
[Amazon Kinesis Data Analytics](https://aws.amazon.com/about-aws/whats-new/2019/11/you-can-now-run-fully-managed-apache-flink-applications-with-apache-kafka/) also provides a way 
```

massdosage commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+[Presto](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-presto.html) that can run Iceberg.
+
+[Amazon Kinesis Data Analytics](https://aws.amazon.com/about-aws/whats-new/2019/11/you-can-now-run-fully-managed-apache-flink-applications-with-apache-kafka/) also provdies a way 
+to run fully managed Apache Flink application. You can include Iceberg in your application Jar and run it in the platform.
```
```suggestion
to run fully managed Apache Flink applications. You can include Iceberg in your application Jar and run it in the platform.
```

jackye1995 commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+
+## Runtime Packages
+
+The first thing to note is that the `iceberg-aws` module is not bundled with any engine runtime.
```
Yes

jackye1995 commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+To use any features described in later sections, you need to include the following packages by yourself:
+
+* the [iceberg AWS package](https://mvnrepository.com/artifact/org.apache.iceberg/iceberg-aws)
+* the [AWS SDK bundle](https://mvnrepository.com/artifact/software.amazon.awssdk/bundle), or individual AWS client packages if you would like to have a minimum dependency footprint. (please note that we use the new AWS v2 SDK instead of v1)
```
Fixed

jackye1995 commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+It is very common for an organization to store all the tables in a single Glue catalog in a single AWS account and run data computation in many different accounts. 
+In this case, you need to specify a Glue catalog ID when initializing `GlueCatalog`.
+The Glue catalog ID you should use is the AWS account ID.
+This is because in each AWS account, there is a single Glue catalog in each AWS region,
+but the region is pre-determined by the Glue web client that is making the call.
+If you would like to access a Glue catalog in a different region, you should configure you AWS client, see more details in [AWS client configuration](#aws-client-configurations).
```
I see, please read if the new version is easier to understand.

jackye1995 commented on site/docs/aws.md:
```diff
@@ -0,0 +1,212 @@
+To enable server side encryption, use the following configuration properties:
+
+* `s3fileio.sse.type`: `none`, `s3`, `kms` or `custom`, default to `none`
+* `s3fileio.sse.key`: a KMS Key ID or ARN for `kms` type (default to `aws/s3`), or a custom base-64 AES256 symmetric key for `custom` type.
```
Yes that is the default at service side.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+
+The `iceberg-aws` module is bundled with Spark and Flink engine runtimes.
+However, the AWS clients are not bundled so that you can use the same version as your application.
+Please note that we use the new AWS v2 SDK instead of v1.
```
How about "You will need to provide the AWS v2 SDK because that is what Iceberg depends on." Saying "instead of v1" makes it sound like you can't use both, so I would omit that as well.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+You can choose to the [AWS SDK bundle](https://mvnrepository.com/artifact/software.amazon.awssdk/bundle), 
+or individual AWS client packages (Glue, S3, DynamoDB, KMS) if you would like to have a minimum dependency footprint.
+
+For example, to use AWS features with Spark 3 and AWS clients version 2.15.40, you can start the SQL shell with:
```
You may want to call out the part that matters here, that this is adding `software.amazon.awssdk:bundle:2.15.40` to the `--packages` argument.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+Iceberg enables the use of [AWS Glue](https://aws.amazon.com/glue) as the `Catalog` implementation.
+When used, an Iceberg `Namespace` is stored as a [Glue Database](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-databases.html), 
+an Iceberg `Table` is stored as a [Glue Table](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html),
+an Iceberg `Snapshot` is stored as a [Glue TableVersion](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html#aws-glue-api-catalog-tables-TableVersion). 
```
Is this true? I thought that only one version of the table was kept and that it pointed to the Iceberg root metadata file. That file contains more than one snapshot.

jackye1995 replied:
Snapshot is probably a bad word, I will just use table version then. 
But Glue also stores all historical table versions for each update.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+
+Iceberg enables the use of [AWS Glue](https://aws.amazon.com/glue) as the `Catalog` implementation.
+When used, an Iceberg `Namespace` is stored as a [Glue Database](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-databases.html), 
+an Iceberg `Table` is stored as a [Glue Table](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html),
```
I would drop references to Iceberg classes here. Most readers are going to care about tables through a SQL or DataFrame interface, not through Iceberg's API. The mapping to Glue database and table names is probably pretty clear.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+an Iceberg `Table` is stored as a [Glue Table](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html),
+an Iceberg `Snapshot` is stored as a [Glue TableVersion](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html#aws-glue-api-catalog-tables-TableVersion). 
+You can start using Glue catalog by specifying the `catalog-impl` as `org.apache.iceberg.aws.glue.GlueCatalog`. 
+More details about loading the catalog can be found in individual engine pages, such as [Spark](../spark/#loading-a-custom-catalog) and [Flink](../flink/#creating-catalogs-and-using-catalogs).
```
I would definitely give a full example of using `GlueCatalog` here, in addition to linking to Spark and Flink pages.

rdblue replied:
Sorry, there's one above. Maybe just point the reader to it.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+You can start using Glue catalog by specifying the `catalog-impl` as `org.apache.iceberg.aws.glue.GlueCatalog`. 
+More details about loading the catalog can be found in individual engine pages, such as [Spark](../spark/#loading-a-custom-catalog) and [Flink](../flink/#creating-catalogs-and-using-catalogs).
+
+### Glue Catalog ID
```
What about adding a table of configuration options instead of sections?

jackye1995 replied:
The reason I am doing this is because the explanation is quite long, which makes the table look very messy.

rdblue replied:
I think it makes sense here, but I would opt for tables in most other places. Part of the appeal is it forces you to be concise.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+
+By default, Glue will store all the table versions created and user can rollback a table to any historical version if needed.
+However, if you are streaming data to Iceberg, this will easily create a lot of Glue table versions.
+Therefore, it is recommended to turn off the archive feature in Glue by setting `gluecatalog.skip-archive` to false.
```
I think it would be a bit cleaner if these catalog options were shorter. this could simply be `skip-archive` in the catalog config. Similarly, the one above could be `catalog-id`.

jackye1995 replied:
This is mostly trying to be consistent with the `s3fileio` properties by having `gluecatalog` as the config namespace. We might be able to make all of them shorter, something like `glue.id` and `s3.staging.dir`. Any thoughts on this? @danielcweeks

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+The lock has the following additional properties:
+
+* `gluecatalog.lock.wait-ms`:  max time to wait for lock acquisition, defaults to 3 minutes
+* `gluecatalog.lock.expire-ms`: max time a table can be locked by a process, defaults to 20 minutes
```
What about `lock-wait-ms` and `lock-expire-ms`?

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+* `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared across all output streams), defaults to the available number of processors in the system
+* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, defaults to 32MB
+* `s3fileio.multipart.threshold`: the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload, defaults to 1.5
+* `s3fileio.staging.dir`: the directory to hold temporary files, defaults to Java's `java.io.tmpdir` property value
```
Great info here, but I think it may be easier to maintain as a table.

jackye1995 replied:
switched to table layout.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,170 @@
+spark-sql --packages org.apache.iceberg:iceberg-spark3-runtime:0.11.0,software.amazon.awssdk:bundle:2.15.40 \
+    --conf spark.sql.catalog.my_catalog=org.apache.iceberg.spark.SparkCatalog \
+    --conf spark.sql.catalog.my_catalog.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog \
+    --conf spark.sql.catalog.my_catalog.warehouse=s3://my-bucket/my-key-prefix \
```
Does the Glue catalog support custom database locations?

jackye1995 replied:
good point, let me add that section

rdblue commented:
@jackye1995 this is great! I noted a few things, but mostly what I would change is organizing the configuration into tables. The text describing all of the features is clear and really helpful.

jackye1995 commented:
@rdblue there are too many tabs currently on Iceberg website, and adding the additional tab for AWS will make the UI ugly. So I moved Flink, Trino and Hive all under an engine's tab, and added a new tab called Integrations that will have topics like AWS, Nessie and JDBC.

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,187 @@
+| Property                          | Default                                            | Description                                            |
+| --------------------------------- | -------------------------------------------------- | ------------------------------------------------------ |
+| s3fileio.multipart.num-threads    | the available number of processors in the system   | number of threads to use for uploading parts to S3 (shared across all output streams)  |
+| s3fileio.multipart.part.size      | 32MB                                               | the size of a single part for multipart upload requests  |
```
Have we released this yet? We would normally make it `s3fileio.multipart.part-size-bytes`:
* `part` is part of "part size" and isn't really part of the hierarchy
* We prefer to have clear units

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,187 @@
+| s3fileio.multipart.num-threads    | the available number of processors in the system   | number of threads to use for uploading parts to S3 (shared across all output streams)  |
+| s3fileio.multipart.part.size      | 32MB                                               | the size of a single part for multipart upload requests  |
+| s3fileio.multipart.threshold      | 1.5                                                | the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload  |
+| s3fileio.staging.dir              | `java.io.tmpdir` property value                    | the directory to hold temporary files  |
```
Is there anything else under staging, or should this be `staging-dir`?

jackye1995 replied:
Looks like there are multiple discussions around the config key names, and these names for `s3fileio` are not designed by me but Daniel. Let me put up another thread for this discussion before release.

jackye1995 replied:
let's use #2050 for the discussion

rdblue commented on site/docs/configuration.md:
```diff
@@ -35,6 +35,20 @@ The properties can be manually constructed or passed in from a compute engine li
+
+| Property                          | Default            | Description                                            |
+| --------------------------------- | ------------------ | ------------------------------------------------------ |
+| lock.impl                         | null               | a custom implementation of the lock manager, the actual interface depends on the catalog used  |
```
The other implementation class properties are `something-impl`. Shouldn't this be `lock-impl` instead to match those? Then the other properties are in the `lock` namespace.

rdblue commented on site/docs/configuration.md:
```diff
@@ -35,6 +35,20 @@ The properties can be manually constructed or passed in from a compute engine li
+| Property                          | Default            | Description                                            |
+| --------------------------------- | ------------------ | ------------------------------------------------------ |
+| lock.impl                         | null               | a custom implementation of the lock manager, the actual interface depends on the catalog used  |
+| lock.table                        | null               | an optional auxiliary table for locking                |
```
Isn't the table required if you're using Dynamo? I would not say it is optional, although it is optional if you're not using Dynamo.

jackye1995 replied:
Yes, this is not a part of AWS documentation, that is why I say it as optional, I can remove that. I will emphasize it is required in the AWS doc.

rdblue replied:
Thanks for pointing that out. You're right that it is optional here. Thanks!

rdblue replied:
You may also want to link to the Dynamo docs if you haven't already, since that's the only implementation currently available.

rdblue commented on site/mkdocs.yml:
```diff
@@ -54,14 +54,16 @@ nav:
-  - Spark:
-    - Getting Started: getting-started.md
+
+  - Engines:
```
The Spark page is getting long enough that I need to break it into separate pages. That's why I reorganized. I like having an `Integrations` tab, though.

I wonder if we can get rid of forward/back instead. And maybe move "About" into the "Project" list.

rdblue commented:
@jackye1995, we can save some space by getting rid of the next/previous links and by moving "About" into "Project":

```diff
diff --git a/site/docs/css/extra.css b/site/docs/css/extra.css
index 4fa7c8036..3d79de02b 100644
--- a/site/docs/css/extra.css
+++ b/site/docs/css/extra.css
@@ -28,6 +28,10 @@
   float: left;
 }
 
+.navbar-right {
+  display: none;
+}
+
 .navbar-brand {
   margin-right: 1em;
 }
diff --git a/site/mkdocs.yml b/site/mkdocs.yml
index 1ce05135b..209495896 100644
--- a/site/mkdocs.yml
+++ b/site/mkdocs.yml
@@ -40,8 +40,8 @@ markdown_extensions:
   - admonition
   - pymdownx.tilde
 nav:
-  - About: index.md
-  - Project:
+  - About:
+    - Project: index.md
     - Community: community.md
     - Releases: releases.md
     - Trademarks: trademarks.md
```

jackye1995 commented:
> @jackye1995, we can save some space by getting rid of the next/previous links and by moving "About" into "Project":

Thanks for the code reference, I just updated the tabs based on that

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,244 @@
+User can choose the ACL level by setting the `s3.acl` property.
+For more details, please read [S3 ACL Documentation](https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html).
+
+### ObjectStoreLocationProvider
```
How about using a description here rather than a class name, like "Object store file layout" or something?

jackye1995 replied:
Yes that's better, let me also remove other class names in titles

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,244 @@
+    data string,
+    category string)
+USING iceberg
+OPTIONS ('location'='s3://my-special-table-bucket')
```
You can also use the `LOCATION` DDL clause

jackye1995 replied:
oh yeah I completely forgot that

rdblue commented on site/docs/aws.md:
```diff
@@ -0,0 +1,244 @@
+Many organizations have customized their way of configuring AWS clients with their own credential provider, access proxy, retry strategy, etc.
+Iceberg allows users to plug in their own implementation of `org.apache.iceberg.aws.AwsClientFactory` by setting the `client.factory` catalog property.
+
+### AssumeRoleAwsClientFactory
```
Is there a better heading for this as well?

jackye1995 replied:
Yes I just updated
